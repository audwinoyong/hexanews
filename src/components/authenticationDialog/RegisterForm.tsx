/**
 * Display register form component.
 */

import * as React from 'react'
import { Dispatch } from 'redux'
import { startCase } from 'lodash'
import { ValidatorForm, TextValidator } from 'react-material-ui-form-validator'
import Button from '@material-ui/core/Button'
import CircularProgress from '@material-ui/core/CircularProgress'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'

import * as actions from './actions'
import { registerFormString } from '../../constants/string'
import styles from './styles'

export type Props = {
  dispatch: Dispatch<any>
  isLoadingRegister: boolean,
  handleCloseDialog: () => void
  onChangeAuthenticationState: () => void
}

export interface ComponentState {
  email: string
  password: string
  name: string
}

class RegisterForm extends React.Component<Props, ComponentState> {

  constructor(props: Props) {
    super(props)
    this.state = {
      email: '',
      password: '',
      name: '',
    }
  }

  disableSubmitButton() {
    const { email, password, name } = this.state
    return (email === '' || password === '' || name === '')
  }

  onRegister = async () => {
    const { email, password, name } = this.state
    const { dispatch, onChangeAuthenticationState } = this.props

    const capitalizedName = startCase(name)
    await dispatch(actions.registerUser(email, password, capitalizedName, onChangeAuthenticationState))
  }

  public render() {
    const { handleCloseDialog, onChangeAuthenticationState, isLoadingRegister } = this.props
    const { email, password, name } = this.state
    return (
      <div>
        <ValidatorForm
          ref='registerForm'
          onSubmit={() => this.onRegister()}
          instantValidate={false}
        >
          <DialogTitle>
            {registerFormString.dialogTitle}
          </DialogTitle>
          <DialogContent>
            <DialogContentText style={styles.descriptionContainer}>
              {registerFormString.dialogDescription}
            </DialogContentText>
            <TextValidator
              label='Email Address'
              onChange={(event: any) => this.setState({ email: event.target.value })}
              name='registerEmail'
              style={styles.descriptionContainer}
              margin='dense'
              fullWidth
              type='email'
              value={email}
              helperText=' '
              validators={['required', 'isEmail']}
              errorMessages={['Please enter email', 'Please enter a valid email']}
            />
            <TextValidator
              label='Password'
              onChange={(event: any) => this.setState({ password: event.target.value })}
              name='registerPassword'
              margin='dense'
              fullWidth
              type='password'
              value={password}
              helperText=' '
              validators={['required', 'matchRegexp:^[a-zA-Z0-9]{6,20}$']}
              errorMessages={['Please enter password',
                'Password length must be between 6-20 characters and contains no special character']}
            />
            <TextValidator
              label='Name'
              onChange={(event: any) => this.setState({ name: event.target.value })}
              name='registerName'
              margin='dense'
              fullWidth
              type='name'
              value={name}
              helperText=' '
              validators={['required', 'minStringLength:3', 'maxStringLength:50',
                'matchRegexp:^[a-zA-Z\\s]+$']}
              errorMessages={['Please enter name',
                'Name field requires a minimum of 3 characters',
                'Name field requires a maximum of 50 characters',
                'Name can only contain alphabetical characters']}
            />
            <DialogContentText style={styles.footerContainer}>
              {registerFormString.haveAccountLabel}
              <span
                onClick={onChangeAuthenticationState}
                style={styles.footerLink}
              >
                {registerFormString.loginHereLabel}
              </span>
            </DialogContentText>
          </DialogContent>
          <DialogActions>
            <Button onClick={handleCloseDialog} color='primary'>
              {registerFormString.cancelButton}
            </Button>
            <Button
              color='primary'
              disabled={this.disableSubmitButton()}
              type='submit'
            >
              {isLoadingRegister ? <CircularProgress size={22} /> : registerFormString.submitButton}
            </Button>
          </DialogActions>
        </ValidatorForm>
      </div>
    )
  }
}

export default RegisterForm