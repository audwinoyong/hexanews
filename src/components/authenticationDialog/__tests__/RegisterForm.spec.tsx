/**
 * Unit tests for checking the behavior of registerForm components
 */

import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import { shallow } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

import RegisterForm, { Props, ComponentState } from '../RegisterForm'

Enzyme.configure({ adapter: new Adapter() })

describe('Register Form', () => {
  it('renders Register Form', () => {
    const component = renderer.create(
      <RegisterForm
        dispatch={jest.fn()}
        onChangeAuthenticationState={jest.fn()}
        isLoadingRegister={false}
        handleCloseDialog={jest.fn()}
        handleOpenRegisterSnackbar={jest.fn()}
        registerError=''
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handle onRegister', () => {
    const dispatch = jest.fn()
    const component = (
      <RegisterForm
        dispatch={dispatch}
        onChangeAuthenticationState={jest.fn()}
        isLoadingRegister={false}
        handleCloseDialog={jest.fn()}
        handleOpenRegisterSnackbar={jest.fn()}
        registerError=''
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: RegisterForm = layout.instance() as RegisterForm

    instance.onRegister()
    layout.update()

    expect(dispatch).toHaveBeenCalled()
  })
})