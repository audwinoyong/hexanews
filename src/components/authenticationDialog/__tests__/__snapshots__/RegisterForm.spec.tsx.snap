// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Form renders Register Form 1`] = `
<div
  className=""
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiDialogTitle-root-98"
    >
      <h2
        className="MuiTypography-root-99 MuiTypography-title-105"
      >
        Register
      </h2>
    </div>
    <div
      className="MuiDialogContent-root-124"
    >
      <p
        className="MuiTypography-root-99 MuiTypography-subheading-106 MuiDialogContentText-root-125"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        Register to access your personalized homepage. Please enter your email address and password.
      </p>
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-126 MuiFormControl-marginDense-128 MuiFormControl-fullWidth-129"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <label
          className="MuiFormLabel-root-135 MuiInputLabel-root-130 MuiInputLabel-formControl-131 MuiInputLabel-animated-134 MuiInputLabel-marginDense-132"
          data-shrink={false}
          htmlFor={undefined}
        >
          Email Address
        </label>
        <div
          className="MuiInput-root-142 MuiInput-fullWidth-149 MuiInput-formControl-143 MuiInput-underline-146"
        >
          <input
            aria-invalid={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-150 MuiInput-inputType-153 MuiInput-inputMarginDense-151"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name="registerEmail"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={false}
            rows={undefined}
            type="email"
            value=""
          />
        </div>
        <p
          className="MuiFormHelperText-root-155 MuiFormHelperText-marginDense-158"
          id={undefined}
        >
           
        </p>
      </div>
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-126 MuiFormControl-marginDense-128 MuiFormControl-fullWidth-129"
      >
        <label
          className="MuiFormLabel-root-135 MuiInputLabel-root-130 MuiInputLabel-formControl-131 MuiInputLabel-animated-134 MuiInputLabel-marginDense-132"
          data-shrink={false}
          htmlFor={undefined}
        >
          Password
        </label>
        <div
          className="MuiInput-root-142 MuiInput-fullWidth-149 MuiInput-formControl-143 MuiInput-underline-146"
        >
          <input
            aria-invalid={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-150 MuiInput-inputType-153 MuiInput-inputMarginDense-151"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name="registerPassword"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={false}
            rows={undefined}
            type="password"
            value=""
          />
        </div>
        <p
          className="MuiFormHelperText-root-155 MuiFormHelperText-marginDense-158"
          id={undefined}
        >
           
        </p>
      </div>
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-126 MuiFormControl-marginDense-128 MuiFormControl-fullWidth-129"
      >
        <label
          className="MuiFormLabel-root-135 MuiInputLabel-root-130 MuiInputLabel-formControl-131 MuiInputLabel-animated-134 MuiInputLabel-marginDense-132"
          data-shrink={false}
          htmlFor={undefined}
        >
          Name
        </label>
        <div
          className="MuiInput-root-142 MuiInput-fullWidth-149 MuiInput-formControl-143 MuiInput-underline-146"
        >
          <input
            aria-invalid={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-150 MuiInput-inputType-153 MuiInput-inputMarginDense-151"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name="registerName"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={false}
            rows={undefined}
            type="name"
            value=""
          />
        </div>
        <p
          className="MuiFormHelperText-root-155 MuiFormHelperText-marginDense-158"
          id={undefined}
        >
           
        </p>
      </div>
      <p
        className="MuiTypography-root-99 MuiTypography-subheading-106 MuiDialogContentText-root-125"
      >
        <p
          className="MuiTypography-root-99 MuiTypography-body1-108"
          style={
            Object {
              "color": "#E53935",
            }
          }
        >
          
        </p>
      </p>
      <p
        className="MuiTypography-root-99 MuiTypography-subheading-106 MuiDialogContentText-root-125"
        style={
          Object {
            "marginTop": 8,
          }
        }
      >
        Have an account? 
        <span
          onClick={[MockFunction]}
          style={
            Object {
              "color": "#03a87c",
              "cursor": "pointer",
            }
          }
        >
          Login here
        </span>
      </p>
    </div>
    <div
      className="MuiDialogActions-root-162"
    >
      <button
        className="MuiButtonBase-root-190 MuiButton-root-164 MuiButton-text-166 MuiButton-textPrimary-167 MuiButton-flat-169 MuiButton-flatPrimary-170 MuiDialogActions-action-163"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-165"
        >
          Cancel
        </span>
        <span
          className="MuiTouchRipple-root-193"
        />
      </button>
      <button
        className="MuiButtonBase-root-190 MuiButtonBase-disabled-191 MuiButton-root-164 MuiButton-text-166 MuiButton-textPrimary-167 MuiButton-flat-169 MuiButton-flatPrimary-170 MuiButton-disabled-184 MuiDialogActions-action-163"
        disabled={true}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
        type="submit"
      >
        <span
          className="MuiButton-label-165"
        >
          Register
        </span>
      </button>
    </div>
  </form>
</div>
`;
